[core] Add authentication to Toolpad Core

bharatkashyap opened the pull request:
Closes https://github.com/mui/mui-toolpad/issues/3419

https://github.com/user-attachments/assets/42993e21-c23c-4746-bcd7-1903d850fece

[Sign In Page docs](https://deploy-preview-3609--mui-toolpad-docs.netlify.app/toolpad/core/react-sign-in-page/)
[Account docs](https://deploy-preview-3609--mui-toolpad-docs.netlify.app/toolpad/core/react-account/)

## To Do List

- [x] Demos and docs for `SignInPage` component
- [x] Demos and docs for ~`User`~  `Account` component
- [x] Add jsdoc to properties for both components to help with vscode intellisense
- [x] Allow customising components through a `slotProps` API
- [ ] Test auth examples functioning with packages from this PR
   - [x] `core-auth-nextjs` working
   - [ ] `core-auth-nextjs-pages` working 

## Follow-up
- [ ] Mention which OAuth Providers are supported
- [ ] `<Account />` demo spacing/layout shift issues
- [ ] Finish the credentials provider:
    - [ ] `RegisterPage` component
    - [ ] `ChangePasswordPage` component
- [ ] Allow customisation through a `slots` API
- [ ] Support multiple layout versions of the `SignInPage`
- [ ] Add email/magic links provider and build an example: https://next-auth.js.org/providers/email

## Open topics
- [ ] `authentication` prop, should it contain the session?
- [ ] Render coloured buttons for OAuth providers needs: https://github.com/mui/material-ui/pull/43003

Janpot commented:
Shaping a bit more the SIgnIn page. For now let's concentrate purely on this component

- clean up implementation
- keep ‘auth’ in the page paths
- start building a pages router version (untested)
- extract SignInPage client component
- integrate branding provider
- make framework agnostic (e.g. remove next.js Image component)
- rename env vars, don’t call them TOOLPAD_…
- start on adding credentials provider support
- fix some spacing issues

Janpot commented:
Further advancements:
- Make full auth flow work in both apps and pages router
- Allow disabling auth for specific pages in pages router, similar to `getLayout` method.
- Add session and authentication contexts
- Add user avatar in dashboard layout
- Lots of fixes

Open questions:
- Naming of `User` component?
- Move `session` prop to `authentication.session` prop?
- Show unauthenticated user avatar with sign in flow when authentication is configured, but no session is present?
- I didn't add tests/docs

apedroferreira commented on playground/nextjs/src/app/auth/error/page.tsx:
```diff
@@ -0,0 +1,62 @@
+'use client';
```
This seems fine, just mentioning that what I did in Studio was to redirect to the signup page when there is an authentication error and show it in a `Snackbar` component.

apedroferreira commented on packages/toolpad-core/src/AppProvider/AppProvider.tsx:
```diff
@@ -50,13 +50,31 @@ export type NavigationItem = NavigationPageItem | NavigationSubheaderItem | Navi
 
+export const SessionContext = React.createContext<Session | null>(null);
+
+export const AutheticationContext = React.createContext<Authentication | null>(null);
```
```suggestion
export const AuthenticationContext = React.createContext<Authentication | null>(null);
```

Janpot commented:
Todays update:

* Add basic docs
* User button becomes login button when no session is available. This enables including it on public pages
* Add basic User component tests
* Fix typos
* Remove sign in button background for now. (If we build custom backgrounds it needs to respect the dark/light theme)

bharatkashyap commented:
> Open questions:
> * Move `session` prop to `authentication.session` prop?

Agreed; also, on wrapping `next-auth` or not - should we eventually wrap all `next-auth` dependencies and re-export them from `@toolpad/core`, or  leave this as it is?

Janpot commented:
toolpad core will be agnostic of authentication library

bharatkashyap commented:
> Open questions:
> 
> * Naming of `User` component?

- Also propose renaming this to `<Account />`, that would support potential additional features we add to this component under the umbrella of account settings and management

bharatkashyap commented:
### Updates

- Fixes to make `credentialsProvider` sign in work
- Allow redirecting to the path provided in the`callbackUrl` search parameter
- Show credential auth errors as notifications on the same page via `useNotifications`
- Refactor some code for the Next.js app router app

Janpot commented on docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:
```diff
@@ -0,0 +1,15 @@
+
+<p class="description">A component that renders a sign in page for your application.</p>
+
+{{"demo": "BasicSignInPage.js"}}
```
This page needs more context. An explanation what is being demoed on the page, why you'd use it, how it interacts with the app provider, how you'd use it to solve different problems...

I propose a general structure:
* In the general demo at the top, that shows off the sign in page, I'd show only 3 oath providers. I'd leave out the credentials provider in this one.
* It needs a section that explains setting it up with an oauth provider. The example here shows only one provider. It also needs to explain which other oauth providers are available to be used with this component.
* It needs a section that explains setting it up with the credentials provider. The example shows the credentials provider only.
* It needs a section on customization. And how it interacts with branding in the App provider. If deep customization is unfinished, add an explanation on how it will work and mark it as being under construction.
* We keep the previous sections agnostic of the auth library, but we add a section on how you'd integrate with next-auth (doesn't necessarily need an embedded demo, but we could link to an example)
* We add a section on how to integrate with next.js pages router and next.js app router (doesn't necessarily need an embedded demo, but we could link to an example)

Janpot commented on docs/data/toolpad/core/components/account/account.md:
```diff
@@ -0,0 +1,17 @@
+components: Account
+---
+
+# Account
```
Needs a bit more context. Explanation about why and how to use this.

Janpot commented on docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:
```diff
@@ -1,7 +1,7 @@
 productId: toolpad-core
 title: Dashboard Layout
-components: AppProvider, DashboardLayout
+components: AppProvider, DashboardLayout, Account
```
Needs a section and demo on the user button. And an explanation on how to customize it (slots?).

bharatkashyap commented:
### Updates

- Added `componentProps` to `SignInPage` to be able to modify props in the `credentials` form components
- Changes to how authentication happens for both playground apps
- Added error notifications upon incorrect credentials
- Added an e2e test to the playground Next.js app to test OAuth and Credentials sign in behaviour
- Added example apps on the lines of the playground apps, and linked to them in the docs
- Misc. cleanup

apedroferreira commented on docs/data/toolpad/core/components/sign-in-page/OAuthSignInPage.tsx.preview:
```diff
@@ -0,0 +1,4 @@
+<SignInPage
```
Would be useful if we could see the list of providers in the preview, maybe by passing them directly or in any other way?

bharatkashyap replied:
Agreed, I'll relook at all the previews to make them more demonstrative

apedroferreira commented on docs/data/toolpad/core/components/sign-in-page/auth.js:
```diff
@@ -0,0 +1,55 @@
+  }),
+  Credentials({
+    credentials: {
+      email: { label: 'Email Address', type: 'email' },
```
Maybe just `Email` is enough?

apedroferreira commented on packages/toolpad-core/src/SignInPage/SignInPage.tsx:
```diff
@@ -0,0 +1,324 @@
+          </Avatar>
+        )}
+
+        <Typography variant="h5" color="textPrimary" gutterBottom>
```
It looks great, I think I'd just try to make the title a bit bolder and bigger by default to see how it looks. Maybe a bit more spacing under the logo too but that's just nitpicking and not sure if needed :D

bharatkashyap replied:
Agreed with the spacing; for the text, I tried different heading levels and `h5` seems the most suitable:
<img width="849" alt="h5" src="https://github.com/user-attachments/assets/a77515bf-2ed5-4d1b-a33d-793d61f977f1">
<img width="851" alt="h4" src="https://github.com/user-attachments/assets/3dc15c3f-8ca0-4c65-9c37-071152c788a1">

apedroferreira replied:
Looks good!

apedroferreira commented on docs/data/toolpad/core/components/sign-in-page/ComponentsPropsSignIn.js:
```diff
@@ -0,0 +1,18 @@
+      componentProps={{
+        email: { autoFocus: false, variant: 'standard' },
+        password: { variant: 'standard' },
+        button: { variant: 'outlined' },
```
`button` seems a bit of a generic name, maybe `submitButton` or something? Also maybe `emailField`, `passwordField`...

apedroferreira commented on docs/data/toolpad/core/components/sign-in-page/BrandingSignInPage.tsx:
```diff
@@ -0,0 +1,50 @@
+  const { mode, systemMode } = useColorSchemeShim();
+  const calculatedMode = (mode === 'system' ? systemMode : mode) ?? 'light';
+  const brandingDesignTokens = getDesignTokens(calculatedMode);
+  const THEME = createTheme({
```
Why can't the default theme just look good here out of the box?

bharatkashyap replied:
Do you mean not adding a custom theme in this demo? I was trying to demonstrate how the sign in page could be styled completely differently with a custom theme in this demo.

apedroferreira replied:
Ah got it, I'm already going to be exemplifying theming in a very basic way in the `AppProvider` component docs page, but I guess it might make sense to have something more specific to show how it would look in the sign up page, so I guess this should be cool.

apedroferreira commented on docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:
```diff
@@ -27,3 +27,9 @@ The `branding` prop in the [AppProvider](https://mui.com/toolpad/core/react-app-
 
 {{"demo": "DashboardLayoutNavigation.js", "height": 640, "iframe": true}}
+
+## Account
```
I don't mind the name `UserAccount` if we used it by the way, seems clearer to me.

bharatkashyap replied:
The issue with components starting with `use` is that the docs builder scripts recognise them as hooks. To me `<Account />` seems okay. We can take up modifying the script in case it seems unusable

apedroferreira replied:
Ok, if that's the main/only reason I guess we can make the part that checks "use" in the scripts be case sensitive? Anyway it's not that important to change this name, only if everyone agrees.

apedroferreira reviewed (changes requested):
Nice work, this is a huge feature! 
I went over everything in general and ended up coming up with lots of feedback - nothing huge that should block the PR besides maybe the docs crashing when "expanding code" as mentioned below, and making all docs pages match in style.
Other than that feel free to address the simpler suggestions and keep the other ones noted for later if they make sense!

**Main issues**

- The Sign In Page component page in the docs has a different background color than the other pages. It looks cool, maybe we could use it in all pages, but all pages should look the same.
- I'm getting an error when trying to "expand code" in the Next Auth docs example, this also kept me from taking a deeper look at them

**Other issues**

- The fonts in the sign in form examples in the docs don't seem to match the default ones in the DashboardLayout. Not sure why that's happening, but they should be the same. (ah, if it's just no theme there we can wrap in an `AppProvider` just to get the default theme there). 
- The "Sign In" button in the top bar doesn't really look like a standard button? I feel like it would look better if it did. The "Sign Out" button in the popup could also have a more standard look, it seems to different from the default buttons in the sign up form to me, for example. I can touch these up later too though.
- In general, the gray buttons look to be in a "disabled" state in a way, maybe they could just use the primary theme color at least for basic buttons?
- Would be cool in my opinion if the buttons for different providers had different colors, and also easier to identify them, but that's not super important. 
- The Credentials example in the docs says "The signIn function will accept a formData parameter in this case." but the `signIn` function in the example is the same as in the example above, so the differences are not shown?
- In Studio we have some detailed instructions on how to configure the different OAuth providers, should we do the same for core?

bharatkashyap commented on examples/core-auth-nextjs-pages/src/pages/auth/signin.tsx:
```diff
@@ -0,0 +1,75 @@
+  return (
+    <SignInPage
+      providers={providers}
+      signIn={async (provider, formData, callbackUrl) => {
```
Agreed, I'll clean this up

bharatkashyap replied:
On further discussion with Jan, brought this back to be inline but with better hygiene for clarity

bharatkashyap commented on examples/core-auth-nextjs/src/app/layout.tsx:
```diff
@@ -0,0 +1,57 @@
+  return (
+    <html lang="en">
+      <body>
+        <SessionProvider session={session}>
```
Agree, that's what I believe as well. The `authentication` prop should contain the `session` object - I was waiting for Jan to return to close this discussion. We can implement this in a follow up.

bharatkashyap commented:
Thanks for the detailed feedback Pedro. Addressed your feedback in the latest commits.
> 
> **Main issues**
> 
> * The Sign In Page component page in the docs has a different background color than the other pages. It looks cool, maybe we could use it in all pages, but all pages should look the same.
- Somehow, not using the `iframe: true` property on the Sign In demos was causing this.
> * I'm getting an error (page crashing) when trying to "expand code" in the Next Auth docs example plus some other later sections.
- This was because of a regression in the monorepo fixed by https://github.com/mui/material-ui/commit/86bc4285f50364cb6d39c769a074a0e02d386ed4; I tried running with the latest version of the monorepo but that one has issues as well. I've downgraded to `6.0.0-alpha.13` where it works fine.
> 
> **Other issues**
> 
> * The fonts in the sign in form examples in the docs don't seem to match the default ones in the DashboardLayout. Not sure why that's happening, but they should be the same. (ah, if it's just no theme there we can wrap in an `AppProvider` just to get the default theme there).
- Also fixed by using `iframe: true`
> * The "Sign In" button in the top bar doesn't really look like a standard button? I feel like it would look better if it did. The "Sign Out" button in the popup could also have a more standard look, it seems too different from the default buttons in the sign up form to me, for example. I can touch these up later too though.
> * In general, the gray buttons look to be in a "disabled" state in a way, maybe they could just use the primary theme color at least for basic buttons?
- Agree, changed the look and feel of the buttons
> * Would be cool in my opinion if the buttons for different providers had different colors, and also easier to identify them, but that's not super important.
- I agree with the differentiation provided by colors; I was earlier using `svg` for logos instead of the Material Icon components; however, that did not integrate too well with the `loading` state of the `LoadingButton`. We can iterate on this in the future, I'll add it as an item open for discussion.
> * The Credentials example in the docs says "The signIn function will accept a formData parameter in this case." but the `signIn` function in the example is the same as in the example above, so the differences are not shown?
- Agreed, I've added some demonstration of the `formData` parameter
> * In Studio we have some detailed instructions on how to configure the different OAuth providers, should we do the same for core? Could be done later, no need to do now.
- I've added some setup instructions inline to the docs page, with a link pointing to Auth.js detailed instructions - similar to the Studio docs.
> * Some examples such as the Next Auth one seem to import important logic from external files, but can that logic actually be seen in the docs examples? Otherwise the user might not be able to understand how to use them
The Next Auth demo doesn't actually perform the sign in action, but I doubt there's a way to make that happen inside the demo. Do you suggest removing the code altogether? For me, we could add some indication that this demo is not supposed to actually work for clarity

apedroferreira commented:
> * The Next Auth demo doesn't actually perform the sign in action, but I doubt there's a way to make that happen inside the demo. Do you suggest removing the code altogether? For me, we could add some indication that this demo is not supposed to actually work for clarity

I see that there's tabs showing the content of every imported file, I guess that works for what I meant!
If it's just an example and doesn't actually work in the docs, and we can't make it work locally somehow, I guess that a comment wouldn't hurt.

apedroferreira commented:
> I've added some setup instructions inline to the docs page, with a link pointing to Auth.js detailed instructions - similar to the Studio docs.

Ok! In retrospect and after today's meeting feel free to make these parts shorter by just linking to the relevant Auth.js documentation and we can provide examples on Github for different providers in time.

apedroferreira reviewed (approved):
As far as my feedback it seems like it's all been addressed, thanks!
I'll approve already on my side even though there were a few other things that were discussed in the last meeting that should probably still be addressed too.

apedroferreira commented on docs/data/toolpad/core/components/sign-in-page/ComponentsPropsSignIn.js:
```diff
@@ -0,0 +1,24 @@
+import * as React from 'react';
+import { AppProvider, SignInPage } from '@toolpad/core';
+
+const providers = [{ id: 'credentials', name: 'Email and Password' }];
```
Oh just one more thing, would be nice to show this provider in the preview for the credentials example too.

bharatkashyap commented on docs/pages/_app.js:
```diff
@@ -216,7 +216,7 @@ function AppWrapper(props) {
           metadata: '',
           name: 'Toolpad Core',
-          versions: [{ text: `v0.0.1`, current: true }],
+          versions: [{ text: `v${toolpadPkgJson.version}`, current: true }],
```
- Remove the hard coded version and use the same version number as Toolpad Studio, with the assumption that we keep version numbers in tandem.

bharatkashyap reviewed (commented):
## Updates

- Changed the error handling flow to not throw errors but to use an `AuthResponse` object
- Consequent changes to the playground and example apps
- Consequent changes to the Sign In page docs
- Removed the Playwright tests for the playground, to take up in a follow up
- Docs improvements for the Account component: use demos only showing the component

apedroferreira commented on docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:
```diff
@@ -0,0 +1,109 @@
+
+{{"demo": "AuthJsSignInApp.js", "iframe": true, "height": 300 }}
+
+#### Setting up
```
I know this goes against my first comment on this, but I think maybe we can skip these detailed setup instructions and instead just link to the Next auth documentation? Or maybe it's actually nice to provide these detailed instructions as an example, but we should make it clearer that users should just follow the Next Auth docs as they're implementing the library directly.
Or maybe these detailed instructions should only be included with the examples themselves and not the component documentation.

bharatkashyap replied:
I can definitely cut down on the instructions, give the link to the Next Auth docs more priority and keep the detailed instructions only in the example.

apedroferreira commented on docs/data/toolpad/core/components/sign-in-page/BrandingSignInPage.js:
```diff
@@ -0,0 +1,53 @@
+  { id: 'credentials', name: 'Email and Password' },
+];
+
+const BRANDING = {
```
Maybe these values being passed should be shown in the preview if that doesn't make the preview too complex or take up too much space.

apedroferreira commented on docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:
```diff
@@ -0,0 +1,109 @@
+
+### 🚧 Layouts
+
+The `SignInPage` component has versions with different layouts for authentication - one column, two column and others such. The APIs of these components is identical. This is in progress.
```
```suggestion
The `SignInPage` component has versions with different layouts for authentication - one column, two column and others such. The APIs of these components are identical. This is in progress.
```

apedroferreira commented on docs/data/toolpad/core/components/account/account.md:
```diff
@@ -0,0 +1,29 @@
+
+<p class="description">A component that renders an account management dropdown for your application.</p>
+
+The `Account` component is a quick and easy way to display an account management menu for authenticated users. It works deeply with the `SignInPage` and `DashboardLayout` components, meaning that it automatically appears to the top navigation bar inside `DashboardLayout` once your users have signed in through the `SignInPage`.
```
```suggestion
The `Account` component is a quick and easy way to display an account management menu for authenticated users. It is deeply integrated with the `SignInPage` and `DashboardLayout` components, meaning that it automatically appears in the top navigation bar inside `DashboardLayout` once your users have signed in through the `SignInPage`.
```

apedroferreira reviewed (approved):
I went mostly through the docs, everything seems good so nothing I wrote should be a major blocking change, I think. Let me know if there are any specific things besides docs that you'd like me to take a look at.

Also one more thing: the forgot password links in the demos are kind of weird as they link to the main MUI home page, maybe we can do something about that?

apedroferreira commented on scripts/docs/buildCoreApiDocs/config/getComponentImports.ts:
```diff
@@ -24,16 +24,16 @@ export function getComponentImports(name: string, filename: string) {
 
     return [
-      `import { ${name} } from '@toolpad-core/${name}';${
+      `import { ${name} } from '@toolpad/core/${name}';${
```
Thanks! Missed these.

Janpot commented on docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:
```diff
@@ -0,0 +1,109 @@
+
+## OAuth
+
+The `SignInPage` component can be set up with an OAuth provider by passing in a list of providers in the `providers` prop, along with a `signIn` function that accepts the `provider` as a parameter.
```
_Can be done in follow-up_

We should also explain which providers are supported

Janpot commented on docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:
```diff
@@ -0,0 +1,109 @@
+The component is composable with any authentication library you might want to use. The following is a functional `SignInPage` with [auth.js](https://authjs.dev/) using GitHub, Next.js App router and server actions.
+
+:::warning
+The following demo does not initiate an actual GitHub authentication flow, since doing that from within an `iframe` is not permitted. Run the [Next.js app directory](https://github.com/mui/mui-toolpad/tree/master/examples/core-auth-nextjs/) example to test this functionality.
```
_follow-up_

This isn't the best form. We should just make it so that it shows a loading spinner for a second or so

bharatkashyap replied:
I've added an image instead, remove the extended instructions and linked to the example that is created as part of this PR

Janpot commented on docs/data/toolpad/core/components/account/AccountDemo.tsx:
```diff
@@ -0,0 +1,36 @@
+  const authentication = React.useMemo(() => {
+    return {
+      signIn: () => {
+        alert('Signing in!');
```
Not a great UX
```suggestion
```

Janpot commented on docs/data/toolpad/core/components/account/account.md:
```diff
@@ -0,0 +1,29 @@
+
+When signed out, the component renders as an inline sign in button within the dashboard layout. If a `session` object is present, the component is rendered as a dropdown containing the user's account details as well as an option to sign out.
+
+{{"demo": "AccountDemo.js", "bg": "gradient" }}
```
_follow up_
The demo has some layout shift when signing in

Janpot commented on docs/data/toolpad/core/components/page-container/PageContent.js:
```diff
@@ -2,7 +2,7 @@ import * as React from 'react';
 import CardContent from '@mui/material/CardContent';
 import Typography from '@mui/material/Typography';
-import Grid from '@mui/material/Unstable_Grid2';
+import Grid from '@mui/material/Grid2';
```
🤔  We may want to keep backwards compatibility with v5.

@mnajdova Will the `/Unstable_Grid2` path be maintained during v6?

bharatkashyap commented:
> Also one more thing: the forgot password links in the demos are kind of weird as they link to the main MUI home page, can we do something about that?

You're right; I've removed them for now - we can add them back once we have those components added?

Janpot reviewed (approved):
Ok, let's merge. We can iterate in follow-up
